http.go

```diff
@@ -1,17 +1,142 @@
+			body        io.ReadCloser
+			err         error
+		)
+		switch enc := r.Header.Get("Content-Encoding"); enc {
+		case "":
+			body = r.Body
+		case "deflate":
+			body, err = zlib.NewReader(r.Body)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				s.logger.WithFields(logrus.Fields{
+					"encoding": enc,
+					"client":   r.RemoteAddr,
+				}).Error("Could not read compressed request body")
+				s.statsd.Count("import.request_error_total", 1, nil, 1.0)
+				return
+			}
+			defer body.Close()
+		default:
+			http.Error(w, enc, http.StatusUnsupportedMediaType)
+			s.logger.WithFields(logrus.Fields{
+				"encoding": enc,
+				"client":   r.RemoteAddr,
+			}).Errorln("Could not determine content-encoding of request")
+			s.statsd.Count("import.request_error_total", 1, nil, 1.0)
+			return
+		}
+
+		if err := json.NewDecoder(body).Decode(&jsonMetrics); err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			s.logger.WithFields(logrus.Fields{
+				logrus.ErrorKey: err,
+				"client":        r.RemoteAddr,
+			}).Error("Could not decode /import request")
+			s.statsd.Count("import.request_error_total", 1, nil, 1.0)
+			return
+		}
+		w.WriteHeader(http.StatusAccepted)
+		// the server usually waits for this to return before finalizing the
+		// response, so this part must be done asynchronously
+		go s.ImportMetrics(jsonMetrics)
+
+		s.statsd.TimeInMilliseconds("import.response_duration_ns", float64(time.Now().Sub(start).Nanoseconds()), []string{"part:request"}, 1.0)
+	})
+
 	return mux
 }
+
+// feed a slice of json metrics to the server's workers
+func (s *Server) ImportMetrics(jsonMetrics []JSONMetric) {
+	start := time.Now()
+
+	// we have a slice of json metrics that we need to divide up across the workers
+	// we don't want to push one metric at a time (too much channel contention
+	// and goroutine switching) and we also don't want to allocate a temp
+	// slice for each worker (which we'll have to append to, therefore lots
+	// of allocations)
+	// instead, we'll compute the fnv hash of every metric in the array,
+	// and sort the array by the hashes
+	sortable := newSortableJSONMetrics(jsonMetrics, len(s.Workers))
+	sort.Sort(sortable)
+
+	startIndex := 0
+	for i := range sortable.metrics {
+		// find the end of the slice belonging to this worker
+		if sortable.hashes[i] != sortable.hashes[startIndex] {
+			// split that slice off, ship it out
+			s.Workers[sortable.hashes[startIndex]].ImportChan <- sortable.metrics[startIndex:i]
+			// and start on the next worker
+			startIndex = i
+		}
+	}
+	// don't forget the last slice
+	if len(sortable.metrics) != 0 {
+		s.Workers[sortable.hashes[startIndex]].ImportChan <- sortable.metrics[startIndex:]
+	}
+
+	s.statsd.TimeInMilliseconds("import.response_duration_ns", float64(time.Now().Sub(start).Nanoseconds()), []string{"part:merge"}, 1.0)
+}
+
+// sorts a set of jsonmetrics by what worker they belong to
+type sortableJSONMetrics struct {
```

Comment:
I've been staring at the `sort` docs for a bit and must admit I can't figure out how it knows to sort this. Clearly you want to sort it by hashes, but then the `metrics` needs to stay in the same order. Could you explain how this works? I'm sure I'm missing something simple…